### ISSUE TITLE
Builds that worked against RC2 failing against master

### ISSUE THREAD

**Issue by pnorman** [mapnik/mapnik#2914]
building node-mapnik 0e90fe8, I get the errors

```
gmake: Entering directory '/usr/home/pnorman/osm/mapnik/kosmtik/node_modules/mapnik/build'
  ACTION binding_gyp_action_before_build_target_generate_protoc_files Release/obj/gen/vector_tile.pb.cc
  ACTION binding_gyp_action_before_build_target_generate_setting Release/obj/gen/mapnik_settings.js
  COPY /usr/home/pnorman/osm/mapnik/kosmtik/node_modules/mapnik/lib/binding/node-v11-freebsd-x64/mapnik_settings.js
  TOUCH Release/obj.target/action_before_build.stamp
  CXX(target) Release/obj.target/mapnik/src/mapnik_logger.o
In file included from ../src/mapnik_logger.cpp:3:
In file included from /home/pnorman/mapnik/include/mapnik/debug.hpp:28:
/home/pnorman/mapnik/include/mapnik/util/singleton.hpp:40:7: error: redefinition of 'CreateUsingNew'
class CreateUsingNew
      ^
/usr/local/include/mapnik/utils.hpp:49:7: note: previous definition is here
class CreateUsingNew
      ^
In file included from ../src/mapnik_logger.cpp:3:
In file included from /home/pnorman/mapnik/include/mapnik/debug.hpp:28:
/home/pnorman/mapnik/include/mapnik/util/singleton.hpp:54:7: error: redefinition of 'CreateStatic'
class CreateStatic
      ^
/usr/local/include/mapnik/utils.hpp:63:7: note: previous definition is here
class CreateStatic
      ^
In file included from ../src/mapnik_logger.cpp:3:
In file included from /home/pnorman/mapnik/include/mapnik/debug.hpp:28:
/home/pnorman/mapnik/include/mapnik/util/singleton.hpp:74:84: error: redefinition of 'singleton'
          template <typename U> class CreatePolicy=CreateStatic> class MAPNIK_DECL singleton
                                                                                   ^
/usr/local/include/mapnik/utils.hpp:100:84: note: previous definition is here
          template <typename U> class CreatePolicy=CreateStatic> class MAPNIK_DECL singleton
                                                                                   ^
In file included from ../src/mapnik_logger.cpp:3:
In file included from /home/pnorman/mapnik/include/mapnik/debug.hpp:28:
/home/pnorman/mapnik/include/mapnik/util/singleton.hpp:132:95: error: redefinition of 'mutex_' with a different type: 'std::mutex' vs 'boost::mutex'
              template <typename U> class CreatePolicy> std::mutex singleton<T,CreatePolicy>::mutex_;
                                                                                              ^
/usr/local/include/mapnik/utils.hpp:168:90: note: previous definition is here
              template <typename U> class CreatePolicy> mutex singleton<T,CreatePolicy>::mutex_;
                                                                                         ^
In file included from ../src/mapnik_logger.cpp:3:
In file included from /home/pnorman/mapnik/include/mapnik/debug.hpp:28:
/home/pnorman/mapnik/include/mapnik/util/singleton.hpp:134:100: error: redefinition of 'pInstance_' with a different type: 'std::atomic<T *>' vs 'T *'
              template <typename U> class CreatePolicy> std::atomic<T*> singleton<T,CreatePolicy>::pInstance_;
                                                                                                   ^
/usr/local/include/mapnik/utils.hpp:172:87: note: previous definition is here
              template <typename U> class CreatePolicy> T* singleton<T,CreatePolicy>::pInstance_=0;
                                                                                      ^
In file included from ../src/mapnik_logger.cpp:3:
In file included from /home/pnorman/mapnik/include/mapnik/debug.hpp:28:
/home/pnorman/mapnik/include/mapnik/util/singleton.hpp:136:89: error: redefinition of 'destroyed_'
              template <typename U> class CreatePolicy> bool singleton<T,CreatePolicy>::destroyed_ = false;
                                                                                        ^
/usr/local/include/mapnik/utils.hpp:174:89: note: previous definition is here
              template <typename U> class CreatePolicy> bool singleton<T,CreatePolicy>::destroyed_=false;
                                                                                        ^
6 errors generated.
mapnik.target.mk:190: recipe for target 'Release/obj.target/mapnik/src/mapnik_logger.o' failed
gmake: *** [Release/obj.target/mapnik/src/mapnik_logger.o] Error 1
```
This used to work. As we're in an API freeze pre-3.0 release, I think this is a bug on the Mapnik side, not the code using Mapnik in node-mapnik.

**Comment by flippmoke** [comment]
Singletons were having some issues that were fixed recently, perhaps thats the issue?

/cc @artemp @springmeyer

**Comment by springmeyer** [comment]
This is a stale install issue and should be fixable by doing `make uninstall && make install` or `rm -rf /usr/local/include/mapnik/ && make install`

**Comment by pnorman** [comment]
`/usr/local/include/mapnik` is my install of the latest release, not a development version.

I'm not sure why it's referencing it at all.

I've done `export PATH=$HOME/mapnik/bin:$PATH` and `which mapnik-config` tells me that it is `/home/pnorman/mapnik/bin/mapnik-config`. `mapnik-config --includes` reports 
`-I/home/pnorman/mapnik/include -I/home/pnorman/mapnik/include/mapnik/agg`.

This worked when I wrote gravitystorm/openstreetmap-carto#1579, so something is behaving differently.

**Comment by danpat** [comment]
For whatever reason, `/usr/local/include` is in your include path first, so `#include <mapnik/utils.hpp>` is pulling in `/usr/local/include/mapnik/utils.hpp` instead of `$HOME/mapnik/include/utils.hpp`.  But it *is* pulling in `$HOME/mapnik/include/util/singleton.hpp`, because that's only found in one place.

The conflict is because of trying to include two different versions of header files.

@sprinmeyer's suggestion should fix the problem.  The "real" fix might be to look at `node-mapnik`'s include path and not have it pull in `/usr/local`

**Comment by pnorman** [comment]
> @springmeyer's suggestion should fix the problem

Removing Mapnik from the system isn't a long term solution, but I tried it and I still get an error, just a different one.

```
gmake: Entering directory '/usr/home/pnorman/osm/mapnik/kosmtik/node_modules/mapnik/build'
  ACTION binding_gyp_action_before_build_target_generate_protoc_files Release/obj/gen/vector_tile.pb.cc
  ACTION binding_gyp_action_before_build_target_generate_setting Release/obj/gen/mapnik_settings.js
  COPY /usr/home/pnorman/osm/mapnik/kosmtik/node_modules/mapnik/lib/binding/node-v11-freebsd-x64/mapnik_settings.js
  TOUCH Release/obj.target/action_before_build.stamp
  CXX(target) Release/obj.target/mapnik/src/mapnik_logger.o
In file included from ../src/mapnik_logger.cpp:1:
../src/utils.hpp:19:10: fatal error: 'mapnik/utils.hpp' file not found
#include <mapnik/utils.hpp>
         ^
1 error generated.
mapnik.target.mk:190: recipe for target 'Release/obj.target/mapnik/src/mapnik_logger.o' failed
gmake: *** [Release/obj.target/mapnik/src/mapnik_logger.o] Error 1
```

Bisecting it, 2eaa90033a897736334cd388b11fcad0f9f9c526 changed the API and should probably be reverted to keep the API freeze in place.

As a horrible ugly hack of a workaround, `ln -s ~/mapnik/include/mapnik/util/singleton.hpp ~/mapnik/include/mapnik/utils.hpp` got me past that compile error and to a different one

```
In file included from ../src/mapnik_map.cpp:9:
In file included from ../node_modules/mapnik-vector-tile/src/vector_tile_processor.hpp:88:
../node_modules/mapnik-vector-tile/src/vector_tile_processor.ipp:35:10: fatal error: 'clipper.hpp' file not found
#include "clipper.hpp"
         ^
3 warnings and 1 error generated.
mapnik.target.mk:190: recipe for target 'Release/obj.target/mapnik/src/mapnik_map.o' failed
```

**Comment by danpat** [comment]
`mapnik-vector-tile` needs to pull down a couple of dependencies at build time (`https://github.com/mapbox/pbf.hpp.git` and `https://github.com/mapnik/clipper.git`), the missing `clipper.hpp` needs to be in `mapnik-vector-tile/deps/clipper`.

These are usually populated by running `make` for `mapnik-vector-tile`, but it looks like that step hasn't been run in your build yet.

**Comment by pnorman** [comment]
Why would these be failing now and not failing before? I'm using the same node-mapnik version, https://github.com/mapnik/node-mapnik/commit/0e90fe8.

**Comment by danpat** [comment]
`clipper.hpp` and `pbf.hpp` are new dependencies for `mapnik-vector-tile` in the last week or so.  This looks like a build bug on either `node-mapnik` or `mapnik`'s npm descriptors (I'm not super familiar with how node modules are packaged).

**Comment by danpat** [comment]
Found it.  It looks like mapnik/node-mapnik@0e90fe8 is is pulling the mapnik-vector-tile tarball from master/HEAD.  This is a moving target, and has recently added some dependencies which are satisfied by `node-mapnik` @ HEAD, but not by the older codebase you're building with.

The latest `node-mapnik` code pulls a versioned npm of `mapnik-vector-tile`, which shouldn't change over time.

@pnorman - how feasible is it for you to build against mapnik/node-mapnik@HEAD rather than against 0e90fe8?  If you need to use 0e90fe8 for some reason, you might need to do some jiggery-pokery to slot an older copy of `mapnik-vector-tile` in there.

TL;DR - two separate issues - build errors when including files from two different versions of `mapnik`, and can no longer build old `node-mapnik` codebase because of mutable dependency.

**Comment by pnorman** [comment]
> @pnorman - how feasible is it for you to build against mapnik/node-mapnik@HEAD rather than against 0e90fe8

I really need a specific version of node-mapnik as HEAD is a moving target. Of course, there's no reason I need *that* version, I just need someone to give me an appropriate git version.

I'm trying f9656f1 and will update accordingly.

**Comment by pnorman** [comment]
f9656f1 of node-mapnik works. The issues seem to be on the node-mapnik side, with using include paths in the wrong order, and the previous node-mapnik version not targeting a specific mapnik-vector-tile version.

**Comment by pnorman** [comment]
> @sprinmeyer's suggestion should fix the problem. The "real" fix might be to look at node-mapnik's include path and not have it pull in /usr/local, or at least have it put mapnik-config paths first in the search order.

Just to note, I built with https://github.com/mapnik/node-mapnik/commit/7b33be0413854006e8c2fabbbdf8464aaa7ffe3f and this bug appears to be fixed.
